File: November LeetCoding Challenge 2021/122. Best Time to Buy and Sell Stock II.cpp



class Solution {
public:
    int maxProfit(vector<int>& a) {
        int n = a.size(), cur0, cur1, prev0, prev1;
        
        cur0 = prev0 = 0;
        cur1 = prev1 = -a[0];
        
        for ( int i = 1; i < n; i++ ) {
            cur0 = max ( prev1 + a[i], prev0 );
            cur1 = max ( prev0 - a[i], prev1 );
            prev0 = cur0;
            prev1 = cur1;
        }
        
        return max ( cur0, cur1 );
    }
};

File: Solution with python/November LeetCoding Challenge 2021/122. Best Time to Buy and Sell Stock II.py
class Solution(object):
    def maxProfit(self, prices):
        cur0 = prev0 = 0
        cur1 = prev1 = -prices[0]
        
        for i in range ( 1, len(prices) ) :
            cur0 = max ( prev1 + prices[i], prev0 )
            cur1 = max ( prev0 - prices[i], prev1 );
            
            prev0 = cur0
            prev1 = cur1
        
        return max ( cur0, cur1 )